zerver/tests/test_message_move_topic.py

```diff
@@ -2281,6 +2291,219 @@ def test_resolved_topic_permissions(self) -> None:
+        )
+
+        result = self.resolve_topic_containing_message(
+            admin_user,
+            id2,
+        )
+        self.assert_json_error(
+            result, "You don't have permission to resolve topics in this channel."
+        )
+
+        result = self.resolve_topic_containing_message(
+            hamlet,
+            id2,
+        )
+        self.assert_json_success(result)
+
+        # Test restrict topics to an anonymous group in a particular channel.
+        original_topic_name = "topic 3"
+        id3 = self.send_stream_message(hamlet, stream.name, topic_name=original_topic_name)
+
+        othello_group_member_dict = UserGroupMembersData(
+            direct_members=[othello.id], direct_subgroups=[]
+        )
+        do_change_stream_group_based_setting(
+            stream, "can_resolve_topics_group", othello_group_member_dict, acting_user=othello
+        )
+
+        result = self.resolve_topic_containing_message(
+            admin_user,
+            id3,
+        )
+        self.assert_json_error(
+            result, "You don't have permission to resolve topics in this channel."
+        )
+        result = self.resolve_topic_containing_message(
+            othello,
+            id3,
+        )
+        self.assert_json_success(result)
+
+        # Test resolving topics while moving topics between different streams.
+        # User can resolve the topic when moving it between streams if they have
+        # permission to resolve topics in either the source or destination stream.
+        original_topic_name = "topic 4"
+        new_stream = self.make_stream("new stream")
+        self.subscribe(hamlet, "new stream")
+        id4 = self.send_stream_message(hamlet, new_stream.name, topic_name=original_topic_name)
+
+        do_change_stream_group_based_setting(
+            new_stream, "can_resolve_topics_group", nobody_group, acting_user=admin_user
+        )
+
+        result = self.api_patch(
+            admin_user,
+            "/api/v1/messages/" + str(id4),
+            {"topic": RESOLVED_TOPIC_PREFIX + original_topic_name, "stream_id": stream.id},
+        )
+        self.assert_json_error(
+            result, "You don't have permission to resolve topics in this channel."
+        )
+
+        # Othello can resolve the topic while moving it since he has permission to resolve topics
+        # in the destination stream.
+        result = self.api_patch(
+            othello,
+            "/api/v1/messages/" + str(id4),
+            {"topic": RESOLVED_TOPIC_PREFIX + original_topic_name, "stream_id": stream.id},
+        )
+        self.assert_json_success(result)
+
+        do_change_realm_permission_group_setting(
+            admin_user.realm,
+            "can_move_messages_between_topics_group",
+            nobody_group,
+            acting_user=None,
+        )
+        original_topic_name = "topic 5"
+        id5 = self.send_stream_message(hamlet, stream.name, topic_name=original_topic_name)
```

Comment:
One of these steps should exercise "unresolved topic" rather than just sending to a new fresh topic.